gc.c

```diff
@@ -6819,7 +6825,27 @@ mark_const_tbl(rb_objspace_t *objspace, struct rb_id_table *tbl)
 #endif
 
 static void each_stack_location(rb_objspace_t *objspace, const rb_execution_context_t *ec,
-                                 const VALUE *stack_start, const VALUE *stack_end, void (*cb)(rb_objspace_t *, VALUE));
+                                 const VALUE *stack_start, const VALUE *stack_end, void *ctx,
+                                 void (*cb)(rb_objspace_t *, void *, VALUE));
+
+static void
+gc_mark_machine_stack_location_maybe(rb_objspace_t *objspace, void *ctx, VALUE obj)
+{
+    gc_mark_maybe(objspace, obj);
+#ifdef RUBY_ASAN_ENABLED
+    rb_execution_context_t *ec = ctx;
```

Comment:
The changes to `each_location` and the callback signature make sense to me. However, It seems a shame to introduce quite a lot of extra machinery in order to support drilling down a pointer that will only ever be used when `RUBY_ASAN_ENABLED`.

Have you looked at whether this has any impact on marking performance? Is there a way we could make this zero overhead if asan is not enabled?

Reply:
I think it would be safe to just stash a pointer to the execution context we're currently marking on `objspace`. That would be zero-cost if ASAN is not enabled, and avoid a lot of the messy diff here. I'll do that instead.